### ISSUE TITLE
CQT implementation disagrees with 1992 algorithm

### ISSUE THREAD

**Issue by KinWaiCheuk** [librosa/librosa#981]
Librosa version: 0.7.0
Operating system: 16.04.5 LTS (Xenial Xerus)
Python version: 3.5.0

I discovered that the CQT obtained by librosa (which uses [2010 algorithm](https://urldefense.proofpoint.com/v2/url?u=https-3A__iem.kug.ac.at_fileadmin_media_iem_projects_2010_smc10-5Fschoerkhuber.pdf&d=DwMFAg&c=slrrB7dE8n7gBJbeO0g-IQ&r=afNfHLTJHZqthcuzp-WCA9L1loeJ5eEh8W9qIC1VN8Y&m=xvnKgouiEDNKuQFBqqbw0FWY8XDHbyXfcjzsktmvLZE&s=spVdTqjQbBpHPJVav87dEsnb7kSSpDOyTp2NtWQe6Wc&e=)) is different from the [1992 algorithm](https://urldefense.proofpoint.com/v2/url?u=https-3A__pdfs.semanticscholar.org_628a_0981e2ed1c33b1b9a88018a01e2f0be0c956.pdf&d=DwMFAg&c=slrrB7dE8n7gBJbeO0g-IQ&r=afNfHLTJHZqthcuzp-WCA9L1loeJ5eEh8W9qIC1VN8Y&m=xvnKgouiEDNKuQFBqqbw0FWY8XDHbyXfcjzsktmvLZE&s=GmYpzavRV-wo3vYhUFWUHzeWi0Na2v5bTd2FW9mqcTM&e=).
Specifically, if I use 6 pure sine waves with same amplitude but different frequency, the librosa algorithm shows that the signal becomes weaker as the frequency goes up. 
While the 1992 algorithm shows similar strength on all frequency. (The results are attached as pictures below).

![CQT doubts](https://user-images.githubusercontent.com/37035802/64352366-7b774d80-d02e-11e9-99fc-350e1245a8ff.png)

Here's the code that I wrote on jupyter notebook.
It compares the CQT obtained by Librosa and the 1992 algorithm.

```python
import numpy as np
from scipy import fft
from scipy.sparse import csr_matrix
from scipy.io.wavfile import read
from librosa.core import note_to_hz
import matplotlib.pyplot as plt
from scipy.signal import get_window
import scipy

from librosa import stft
from cqt import cqt
from resampy import resample

# parameters
# nargin = 6
bins = 24 # bins per octave
fs = 22050 # sampling rate

# fmax = 61.73541266 # same as 1 octave
fmax = 880*2
fmin = note_to_hz('C1')
thresh= 0.0054

s = np.linspace(0,1,fs)
x = np.sin(2*np.pi*50*s, dtype=np.float32) + np.sin(2*np.pi*110*s, dtype=np.float32) \
    + np.sin(2*np.pi*220*s, dtype=np.float32) + np.sin(2*np.pi*440*s, dtype=np.float32) \
    + np.sin(2*np.pi*880*s, dtype=np.float32) + np.sin(2*np.pi*1600*s, dtype=np.float32) 

plt.plot(x[:1000])
plt.title('Input Signal', size=20)

def nextpow2(A):
    return int(np.ceil(np.log2(A)))

def create_cqt_kernals(fmin=55, n_bins=162, fmax=None, bins=24, fs=22050, window='hann'):
    """
    '''Compute the constant-Q transform kenrals.

    This implementation is based on Brown 1992 paper [1].

    .. [1] Schoerkhuber, Christian, and Anssi Klapuri.
        "An efficient algorithm for the calculation of a constant Q transform"
        1992.
    """
    
    Q = 1/(2**(1/bins)-1)
    if fmax != None:
        K = np.ceil(bins * np.log2(fmax / fmin)) # Calculate the number of bins
    else:
        K = n_bins
        fmax = fmin*2**(K/bins)
    fftLen = 2**nextpow2(np.ceil(Q * fs / fmin))
    minWin = 2**nextpow2(np.ceil(Q * fs / fmax))
    tempKernel = np.zeros((int(K), int(fftLen)), dtype=np.complex64)
    specKernel = np.zeros((int(K), int(fftLen)), dtype=np.complex64)
    freqs = fmin * 2.0 ** (np.r_[0:K] / np.float(bins))
    for k in range(0, int(K)):
        freq = freqs[k]
        l = np.ceil(Q * fs / freq)
        if l%2==1: # pad more zeros on RHS
            start = int(np.ceil(fftLen / 2.0 - l / 2.0))-1
        else:
            start = int(np.ceil(fftLen / 2.0 - l / 2.0))
        tempKernel[k, start:start + int(l)] = get_window(window,int(l), fftbins=True)*np.exp(np.r_[-l//2:l//2]*1j*2*np.pi*freq/fs)/l
        # specKernel[k, :]=fft(conj(tempKernel[k, :]))
        specKernel[k, :] = fft(tempKernel[k])
        
    return specKernel[:,:fftLen//2+1], fftLen

# Calculating CQT (1992 Approach)

cqt_filter, kernal_width = create_cqt_kernals(fmin)
cqt_filter_abs = abs(cqt_filter)

x_fft = stft(x, kernal_width, hop_length=512, window='ones')
CQT = np.dot(cqt_filter, x_fft)
plt.imshow(abs(CQT),aspect='auto', origin='lower')

# Librosa CQT

from librosa.core import cqt

cqt_librosa = cqt(x, fmin=fmin, n_bins=CQT.shape[0], bins_per_octave=bins, norm=None)
plt.imshow(abs(cqt_librosa),aspect='auto', origin='lower')

# Look at only 1 time slice

plt.figure(figsize=(14,5))
plt.plot(abs(CQT[:,10]))
plt.plot(abs(cqt_librosa[:,10]), 'r--')
plt.xlabel('CQT bins', size=16)
plt.title('CQT at time slice 11', size=20)
plt.tick_params(labelsize=16)

plt.legend(['1992 Algorithm', 'Librosa'])
```

**Comment by bmcfee** [comment]
With a bit of editing, I was able to run your example code and reproduce your figure, though I'm not sure I 100% understand your implementation yet.

A couple of things though: you're calling librosa.cqt() with `norm=None`, meaning that filters are unnormalized.  This will skew the results depending on the window function, and should not be expected to produce a flat response.  Remember that this normalization is applied to each filter independently.

It otherwise looks to me like the discrepancy between your implementation and ours is a factor of `sqrt(filter_length)` (for each basis) when `scale=True` (default), or a factor of `filter_length` when `scale=False`.

Here are some modifications to your code that produces a flat response.  I've also disabled as many speed optimizations as possible to ensure proper comparison (sparsity and fast resampling).

```python
cqt_librosa = cqt(x, fmin=fmin, n_bins=CQT.shape[0], bins_per_octave=bins,
                         norm=1, scale=False, res_type='fft', sparsity=0)

lengths = librosa.filters.constant_q_lengths(fs, fmin, n_bins=len(cqt_librosa),
                                                                     bins_per_octave=bins)

# Look at only 1 time slice

plt.figure(figsize=(14,5))
plt.plot(abs(CQT[:,len(x)//2 // 512]))
plt.xlabel('CQT bins', size=16)
plt.title('CQT at time slice 11', size=20)
plt.tick_params(labelsize=16)
plt.legend(['1992 Algorithm'])

plt.figure(figsize=(14,5))
plt.plot(abs(cqt_librosa[:,len(x)//2 // 512]) / lengths, 'r--')
plt.xlabel('CQT bins', size=16)
plt.title('CQT at time slice 11', size=20)
plt.tick_params(labelsize=16)

plt.legend(['Librosa'])
```
![image](https://user-images.githubusercontent.com/1190540/64430629-a5f4f500-d086-11e9-9bcc-ae9887d24244.png)
![image](https://user-images.githubusercontent.com/1190540/64430672-c1600000-d086-11e9-8fda-705697901b32.png)

There also appears to be a factor of FFT-length in your implementation, which we normalize out explicitly to maintain scale when the FFT size changes due to different frequency ranges in use.

The sqrt(len) scaling was included in librosa to maintain smooth responses across octaves when resampling, and also to maintain consistency with the pseudo-cqt when using hybrid mode.  All in all, I think this is expected behavior, and it's easy enough to resolve against the 92 algorithm if you don't want this normalization in place.

**Comment by lostanlen** [comment]
Related: #341, which was closed by #347

**Comment by KinWaiCheuk** [comment]
In `librosa.filters`, line 550, you normalize the filter with L1 normalization. Then, in the `constant.py`, there’s a function called `__cqt_filter_fft`. Here, you normalize it again with `basis *= lengths[:, np.newaxis] / float(n_fft)` at line 755.
So, the mathematical formula for your cqt filters looks like   
<img src="https://user-images.githubusercontent.com/37035802/64663739-53994700-d47f-11e9-8c86-8c7b54674c45.png" width="100" height="50">

In the 1992 paper, they have shown all the necessary steps to derive the normalization for the filters.
  
![image](https://user-images.githubusercontent.com/37035802/64663999-429d0580-d480-11e9-9380-0bedf0340fca.png)

In other words, they have proved that the kernel normalization should be
<img src="https://user-images.githubusercontent.com/37035802/64664030-64968800-d480-11e9-84de-690cfa6b8736.png" width="70" height="50">.

After running your modification, I can get a similar shape. But the magnitude is still different with a factor 35000. I can't quite understand the maths behind the normalization in your implementation. Do you have the mathematical steps explaining how to arrive this kind of normalization?

**Comment by bmcfee** [comment]
I think your notation is a little confused here: their `N` is our `n_fft`, so it would be a bit more accurate to say that our normalization looks like `(K / N) * (L / |K|_1)`, where `L` is the length (samples) of the time-domain filter `K` -- `N[k_cq]` in their notation.

> In other words, they have proved that the kernel normalization should be

I disagree with this interpretation.  They've chosen one particular normalization, but it will give very different results depending on the choice of window function.  Brown and Puckette only used a hamming window, but our implementation allows for arbitrary window functions.  As such, filters of the same length will have different (L1) norms, so you can think of `L / ||K||_1` as a calibration factor that compensates for gain induced by the choice of window.

**Comment by lostanlen** [comment]
Thank you @KinWaiCheuk for raising this issue and thanks @bmcfee for explaining the rationale behind the current behavior. I think that this issue can be closed without edit to the cqt code.
